packages/as-transformer/src/helpers/protobuf.ts

```diff
@@ -0,0 +1,125 @@
+enum ProtoType {
+  Double = 'double',
+  Float = 'float',
+  Int32 = 'int32',
+  Int64 = 'int64',
+  UInt32 = 'uint32',
+  UInt64 = 'uint64',
+  SInt32 = 'sint32',
+  SInt64 = 'sint64',
+  Fixed32 = 'fixed32',
+  Fixed64 = 'fixed64',
+  SFixed32 = 'sfixed32',
+  SFixed64 = 'sfixed64',
+  Bool = 'bool',
+  String = 'string',
+  Bytes = 'bytes',
+}
+
+interface FieldSpec {
+  type: ProtoType;
+  repeated: boolean;
+}
+
+export interface Argument {
+  name: string;
+  type: string;
+}
+
+export function generateProtoFile(
+  name: string,
+  args: Argument[],
+  returnedType: string,
+): string {
+  const argumentMessages = args.map((arg, index) =>
+    generateArgumentMessage(arg, index + 1),
+  );
+  const fields = argumentMessages.join('\n');
+
+  let protoFile = `syntax = "proto3";
+
+message ${name} {
+${fields}
+}`;
+
+  if (returnedType != '' && returnedType != 'null') {
+    const argumentResponse: Argument = {
+      name: 'value',
+      type: returnedType,
+    };
+
+    const response = generateArgumentMessage(argumentResponse, 1);
+
+    protoFile += `
+
+message ${name}Response {
+${response}
+}`;
+  }
+
+  return protoFile;
+}
+
+function generateArgumentMessage(arg: Argument, index: number): string {
+  const fieldName = arg.name;
+  const fieldSpec = getTypeName(arg.type);
+  const fieldType = (fieldSpec.repeated ? 'repeated ' : '') + fieldSpec.type;
+  return `  ${fieldType} ${fieldName} = ${index};`;
+}
+
+function getTypeName(type: string): FieldSpec {
+  let spec: FieldSpec = {
+    type: ProtoType.Int32,
+    repeated: false,
+  };
+
+  switch (type) {
+    case 'bool':
+      spec.type = ProtoType.Bool;
+      break;
+    case 'i8':
+    case 'Int8Array':
+    case 'i16':
+    case 'Int16Array':
+    case 'i32':
+    case 'Int32Array':
+      spec.type = ProtoType.Int32;
+      break;
+    case 'i64':
+    case 'Int64Array':
+    case 'isize':
+      spec.type = ProtoType.Int64;
+      break;
+    case 'u8':
+    case 'Uint8Array':
+    case 'u16':
+    case 'Uint16Array':
+    case 'u32':
+    case 'Uint32Array':
+      spec.type = ProtoType.UInt32;
+      break;
+    case 'u64':
+    case 'Uint64Array':
+    case 'usize':
+      spec.type = ProtoType.UInt64;
+      break;
+    case 'f32':
+    case 'Float32Array':
+      spec.type = ProtoType.Float;
+      break;
+    case 'f64':
+    case 'Float64Array':
+      spec.type = ProtoType.Double;
+      break;
+    case 'string':
+    case 'Array<string>':
+      spec.type = ProtoType.String;
+      break;
+    default:
+      throw new Error(`Unsupported type: ${type}`);
+  }
```

Comment:
I would prefer an object instead of switch

Reply:
Can you suggest some code ? I don't understand ...

Reply:
of course (just a preference for the readability):
```typescript
function getTypeName(type: string): FieldSpec {
  let spec: FieldSpec = {
    type: ProtoType.Int32,
    repeated: false,
  };

  spec.type = typeMap[type] ?? throwError(`Unsupported type: ${type}`);
  spec.repeated = type.includes('Array');

  return spec;
}

const typeMap: Record<string, ProtoType> = {
  'bool': ProtoType.Bool,
  'i8': ProtoType.Int32,
  'Int8Array': ProtoType.Int32,
  'i16': ProtoType.Int32,
  'Int16Array': ProtoType.Int32,
  'i32': ProtoType.Int32,
  'Int32Array': ProtoType.Int32,
  'i64': ProtoType.Int64,
  'Int64Array': ProtoType.Int64,
  'isize': ProtoType.Int64,
  'u8': ProtoType.UInt32,
  'Uint8Array': ProtoType.UInt32,
  'u16': ProtoType.UInt32,
  'Uint16Array': ProtoType.UInt32,
  'u32': ProtoType.UInt32,
  'Uint32Array': ProtoType.UInt32,
  'u64': ProtoType.UInt64,
  'Uint64Array': ProtoType.UInt64,
  'usize': ProtoType.UInt64,
  'f32': ProtoType.Float,
  'Float32Array': ProtoType.Float,
  'f64': ProtoType.Double,
  'Float64Array': ProtoType.Double,
  'string': ProtoType.String,
  'Array<string>': ProtoType.String,
};

function throwError(message: string): never {
  throw new Error(message);
}
```

Reply:
hum, I will keep my switch :rofl: